### PR TITLE
Topic add gaussian profile

### PR DESCRIPTION
**Pull request by BrianMarre**
replaces PR #4973

- [x] requires PR #4982 to be merged first

### CODE REVIEW

**Comment by BrianMarre** [comment]
@steindev ready for review

File: lib/python/picongpu/picmi/distribution/Distribution.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+class Distribution(pydantic.BaseModel):
+    rms_velocity: typing.Tuple[float, float, float] = (0, 0, 0)
+    """thermal velocity spread [m/s]"""
+
+    directed_velocity: typing.Tuple[float, float, float] = (0, 0, 0)
+    """Directed, average, proper velocity [m/s]"""
+
+    fill_in: bool = True
+    """Flags whether to fill in the empty spaced opened up when the grid moves"""
```
Comment (steindev) [inline lib/python/picongpu/picmi/distribution/Distribution.py]: spaced -> space

File: lib/python/picongpu/picmi/distribution/GaussianDistribution.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    Will create the following profile:
+    - for y < center_front:                density * exp(factor * abs(y - center_front/sigma_front)**power)
+    - for center_front <= y <=center_rear: density
+    - for y > center_rear:                 density * exp(factor * abs(y - center_rear/sigma_rear)**power)
+
+    with y being the position in the simulation box
+    """
+
+    density: float
+    """density particle number density, [m^-3]"""
```
Comment (steindev) [inline lib/python/picongpu/picmi/distribution/GaussianDistribution.py]: remove first density

File: lib/python/picongpu/picmi/distribution/GaussianDistribution.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    center_front: float
+    """center_front center of gaussian ramp at the front, [m]"""
```
Comment (steindev) [inline lib/python/picongpu/picmi/distribution/GaussianDistribution.py]: remove `center_front`

File: lib/python/picongpu/picmi/distribution/GaussianDistribution.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    center_rear: float
+    """center_rear center of the gaussian ramp at the rear, [m]"""
```
Comment (steindev) [inline lib/python/picongpu/picmi/distribution/GaussianDistribution.py]: remove `center_rear`

File: lib/python/picongpu/picmi/distribution/GaussianDistribution.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    sigma_front: float
+    """sigma of the gaussian ramp at the front, [m]"""
+    sigma_rear: float
+    """sigma of the gaussian ramp at the front, [m]"""
```
Comment (steindev) [inline lib/python/picongpu/picmi/distribution/GaussianDistribution.py]: front -> rear

File: lib/python/picongpu/pypicongpu/species/operation/densityprofile/gaussian.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    density( gasCenterFront >= y >= gasCenterRear ) = density_si
+    density( gasCenterRear < y )                    =
+        density_si * exp(gasFactor * (abs( (y - gasCenterRear) / gasSigmaRear))^gasPower)
+    """
+
+    gas_center_front = util.build_typesafe_property(float)
+    """position of the front edge of the constant of middle the density profile in meter"""
+
+    gas_center_rear = util.build_typesafe_property(float)
+    """position of the rear edge of the constant of middle the density profile in meter"""
```
Comment (steindev) [inline lib/python/picongpu/pypicongpu/species/operation/densityprofile/gaussian.py]: this is broken

File: lib/python/picongpu/pypicongpu/species/operation/densityprofile/gaussian.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+    density( y < gasCenterFront )                   =
+        density_si * exp(gasFactor * (abs( (y - gasCenterFront) / gasSigmaFront))^gasPower)
+    density( gasCenterFront >= y >= gasCenterRear ) = density_si
+    density( gasCenterRear < y )                    =
+        density_si * exp(gasFactor * (abs( (y - gasCenterRear) / gasSigmaRear))^gasPower)
+    """
+
+    gas_center_front = util.build_typesafe_property(float)
+    """position of the front edge of the constant of middle the density profile in meter"""
```
Comment (steindev) [inline lib/python/picongpu/pypicongpu/species/operation/densityprofile/gaussian.py]: also broken

File: lib/python/picongpu/pypicongpu/species/operation/densityprofile/gaussian.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+from typeguard import typechecked
+
+
+@typechecked
+class Gaussian(DensityProfile):
+    """
+    gaussian density profile
+
+    density( y < gasCenterFront )                   =
+        density_si * exp(gasFactor * (abs( (y - gasCenterFront) / gasSigmaFront))^gasPower)
```
Comment (steindev) [inline lib/python/picongpu/pypicongpu/species/operation/densityprofile/gaussian.py]: density_si -> density, also in the line below

File: lib/python/picongpu/pypicongpu/species/operation/densityprofile/gaussian.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
+"""
+This file is part of the PIConGPU.
+Copyright 2024 PIConGPU contributors
+Authors: Brian Edward Marre, Afshari Masoud
```
Comment (steindev) [inline lib/python/picongpu/pypicongpu/species/operation/densityprofile/gaussian.py]: change order of Masoud's names

File: lib/python/picongpu/pypicongpu/species/operation/densityprofile/gaussian.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    """
+
+    gas_center_front = util.build_typesafe_property(float)
+    """position of the front edge of the constant of middle the density profile in meter"""
+
+    gas_center_rear = util.build_typesafe_property(float)
+    """position of the rear edge of the constant of middle the density profile in meter"""
+
+    gas_sigma_front = util.build_typesafe_property(float)
+    """distance in away from gasCenterFront until the gas density decreases to its 1/e-th part in meter"""
```
Comment (steindev) [inline lib/python/picongpu/pypicongpu/species/operation/densityprofile/gaussian.py]: broken

File: test/python/picongpu/quick/picmi/distribution.py [hunk 11]
```diff
@@ -448,3 +448,84 @@ def test_drift(self):
         self.assertAlmostEqual(drift.direction_normalized[0], 0.9370354841199405)
         self.assertAlmostEqual(drift.direction_normalized[1], 0.34920746753855203)
         self.assertAlmostEqual(drift.direction_normalized[2], 0.004318114799291135)
+
+
+class TestPicmiGaussianDistribution(unittest.TestCase, HelperTestPicmiBoundaries):
+    def _get_distribution(self, lower_bound=[None, None, None], upper_bound=[None, None, None]):
+        return picmi.GaussianDistribution(
+            density=42.42,
```
Comment (steindev) [inline test/python/picongpu/quick/picmi/distribution.py]: Define as global variable (or so) in order to ensure testing against the same values later on.

File: test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py [hunk 12]
```diff
@@ -0,0 +1,283 @@
+"""
+This file is part of the PIConGPU.
+Copyright 2021-2023 PIConGPU contributors
```
Comment (steindev) [inline test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py]: 2024

File: test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+
+from picongpu.pypicongpu.species.operation.densityprofile import Gaussian, DensityProfile
+
+import unittest
+import typeguard
+
+
+class TestGaussian(unittest.TestCase):
+    def test_inheritance(self):
+        """foil is a density profile"""
```
Comment (steindev) [inline test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py]: gaussian?

File: test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+        self.assertTrue(isinstance(Gaussian(), DensityProfile))
+
+    def test_basic(self):
+        """simple scenario works, other plasma ramps are tested separately"""
```
Comment (steindev) [inline test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py]: update?

File: test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+        g = Gaussian()
+
+        g.gas_center_front = 1.0
```
Comment (steindev) [inline test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py]: Define global (or so) variables to ensure testing against equal values.

File: test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+        # invalid gas_center_rear
+        for invalid in [-1.0]:
+            # assignment passes, but check catches the error
+            g.gas_center_rear = invalid
+            with self.assertRaisesRegex(ValueError, ".*gas_center_rear.* >= 0.*"):
+                g.check()
+
+        # rear < front
+        # assignment passes, but check catches the error
+        g.gas_center_rear = 0.5
```
Comment (steindev) [inline test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py]: put `0.9*gas_center_front`

File: test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+        # invalid gas_center_front
+        for invalid in [-1.0]:
+            # assignment passes, but check catches the error
+            g.gas_center_front = invalid
+            with self.assertRaisesRegex(ValueError, ".*gas_center_front.* >= 0.*"):
+                g.check()
+
+        # front > rear
+        # assignment passes, but check catches the error
+        g.gas_center_front = 3.0
```
Comment (steindev) [inline test/python/picongpu/quick/pypicongpu/species/operation/densityprofile/gaussian.py]: along the lines of above comment, use 1.1*...

**Review by steindev — CHANGES_REQUESTED** [review]

**Comment by steindev** [comment]
@BrianMarre plkease be aware that this branch currently has conflicts, please resolve these too before your next push.

**Comment by BrianMarre** [comment]
@steindev ready for review

**Review by steindev — APPROVED** [review]

**Pull request merged**